packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import invariant from 'shared/invariant';
+import {
+  enableInteractionTracking,
+  enableInteractionTrackingObserver,
+} from 'shared/ReactFeatureFlags';
+import now from './InteractionTrackingNow';
+
+export type Interaction = {|
+  __count?: number,
+  id: number,
+  name: string,
+  timestamp: number,
+|};
+
+export type Subscriber = {
+  // A new interaction has been created via the track() method.
+  onInteractionTracked: (interaction: Interaction) => void,
+
+  // All scheduled async work for an interaction has finished.
+  onInteractionScheduledWorkCompleted: (interaction: Interaction) => void,
+
+  // New async work has been scheduled for a set of interactions.
+  // When this work is later run, onWorkStarted/onWorkStopped will be called.
+  // A batch of async/yieldy work may be scheduled multiple times before completing.
+  // In that case, onWorkScheduled may be called more than once before onWorkStopped.
+  // Work is scheduled by a "thread" which is identified by a unique ID.
+  onWorkScheduled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of scheduled work has been canceled.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkCanceled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has started for a set of interactions.
+  // When this work is complete, onWorkStopped will be called.
+  // Work is not always completed synchronously; yielding may occur in between.
+  // A batch of async/yieldy work may also be re-started before completing.
+  // In that case, onWorkStarted may be called more than once before onWorkStopped.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type SubscriberRef = {
+  current: Subscriber | null,
+};
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
```

Comment:
Interesting.

Using "any" to mask a null initialization value seems equally potentially dangerous but... I don't mind removing all of the Flow casts 😅

Reply:
Both cases has the danger of us using a field while not in the feature flag branch.

In the previous approach, every time you use it  risk typing `(interactionsRef: any)` instead of `((interactionsRef: any): InteractionsRef)` which would cover up additional errors.

You also risk typing `((subscriberRef: any): InteractionsRef)` which is just the wrong type.

The number of times you have to use `any` increases the risk of any of these types of additional mistakes.

When it is in a single location you only risk making this mistake once, but that's easy to review, and since you're using `null` as the initial value the risk of typing `subscriberRef` instead of `null` seems pretty small.